backend/clubs/views.py

```diff
@@ -4778,47 +4817,95 @@ def checkout_success_callback(self, request, *args, **kwargs):
                            properties:
                                 detail:
                                     type: string
+                                success:
+                                    type: boolean
         ---
         """
+        tt = request.data.get("transient_token")
         cart = get_object_or_404(Cart, owner=self.request.user)
 
-        for ticket in cart.tickets.select_for_update().all():
-            ticket.owner = request.user
-            ticket.carts.clear()
-            # ticket.send_confirmation_email()
-            ticket.save()
+        try:
+            _, status, transaction_data = TransientTokenDataApi(
+                settings.CYBERSOURCE_CONFIG
+            ).get_transaction_for_transient_token(tt)
 
-        return Response({"detail": "callback successful"})
+            if not transaction_data or (status % 100) in [4, 5]:
+                raise ApiException(
+                    reason=f"Received {transaction_data} with HTTP status {status}"
+                )
+            transaction_data = json.loads(transaction_data)
 
-    @action(detail=False, methods=["post"])
-    @transaction.atomic
-    def buy(self, request, *args, **kwargs):
-        """
-        Buy held tickets in a cart
-        ---
-        requestBody:
-            content: {}
-        responses:
-            "200":
-                content:
-                    application/json:
-                        schema:
-                            allOf:
-                                - $ref: "#/components/schemas/Ticket"
-        ---
-        """
+        except ApiException as e:
+            # Cleanup state since the purchase failed
+            for ticket in cart.tickets.select_for_update().all():
+                ticket.holder = None
```

Comment:
My thinking:

if `initiate_checkout` fails -- card details entered incorrectly or something of that sort... user error (HTTP 400). Don't clear holds and give another chance to the user.

if `complete_checkout` fails -- something very bad happened; we got a transient token but we couldn't auth using it. (HTTP 500). We have a dangling transient token now: maybe the service will fail to checkout again if we input the card details, or it's down entirely. Either way, it's out of our control and we should clear holds...

I'll add status codes to responses make this clear. I should have done that to begin with :)

Reply:
Sorry I might be misunderstanding, but isn't `initiate_checkout` completely independent of the user's card details, since we just take their cart total and generate a capture context? 

In `complete_checkout`, if the user's entered their card details incorrectly, then the authorization will fail and that transient token is now useless. Even if the authorization is successful, the capture could still fail if the bank flags the transaction as fraudulent.

However I'm thinking it might be better to maintain the hold and send them back to the cart page so they can restart checkout with new payment details.

Reply:
Resolved offline

Reply:
Resolved offline:

Transient tokens represent a complete payment authorization after validating payment details and funds in someone's account for a charge. So, if a payment fails after getting a transient token, then something truly went wrong...which means nuking holds is expected.